cluster_experiments/analysis_plan.py

```diff
@@ -0,0 +1,264 @@
+        ----------
+        tests : List[HypothesisTest]
+            A list of HypothesisTest instances
+        variants : List[Variant]
+            A list of Variant instances
+        variant_col : str
+            The name of the column containing the variant names.
+        alpha : float
+            significance level used to construct confidence intervals
+        """
+        self._validate_inputs(tests, variants, variant_col)
+        self.tests = tests
+        self.variants = variants
+        self.variant_col = variant_col
+        self.alpha = alpha
+
+    @staticmethod
+    def _validate_inputs(
+        tests: List[HypothesisTest], variants: List[Variant], variant_col: str
+    ):
+        """
+        Validates the inputs for the AnalysisPlan class.
+
+        Parameters
+        ----------
+        tests : List[HypothesisTest]
+            A list of HypothesisTest instances
+        variants : List[Variant]
+            A list of Variant instances
+        variant_col : str
+            The name of the column containing the variant names.
+
+        Raises
+        ------
+        TypeError
+            If tests is not a list of HypothesisTest instances or if variants is not a list of Variant instances.
+        ValueError
+            If tests or variants are empty lists.
+        """
+        if not isinstance(tests, list) or not all(
+            isinstance(test, HypothesisTest) for test in tests
+        ):
+            raise TypeError("Tests must be a list of HypothesisTest instances")
+        if not isinstance(variants, list) or not all(
+            isinstance(variant, Variant) for variant in variants
+        ):
+            raise TypeError("Variants must be a list of Variant instances")
+        if not isinstance(variant_col, str):
+            raise TypeError("Variant_col must be a string")
+        if not tests:
+            raise ValueError("Tests list cannot be empty")
+        if not variants:
+            raise ValueError("Variants list cannot be empty")
+
+    def analyze(
+        self,
+        exp_data: pd.DataFrame,  # , pre_exp_data: Optional[pd.DataFrame], alpha=0.05
+    ) -> DataFrame:
+        ...
+        # add methods to prepare the filtered dataset based on variants and slicers
+        # add methods to run the analysis for each of the hypothesis tests, given a filtered dataset
+        # store each row as a HypothesisTestResults object
+        # wrap all results in an AnalysisPlanResults object
+
+        # -----
+
+        # add all kind of checks on the inputs at the beginning using the data structures
+        # todo: ...
+        # do it before running the computations below
+
+        test_results = []
+        treatment_variants: List[Variant] = self.get_treatment_variants()
+        control_variant: Variant = self.get_control_variant()
+
+        for test in self.tests:
+            # todo: add cupac handler here
+            analysis_class = analysis_mapping[test.analysis_type]
```

Comment:
analysis plan has a lot of logic here and hypothesis test isn't doing much. what can we move from here to hypothesis test?